provider/aws_sd.go

```diff
@@ -0,0 +1,626 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"strings"
+
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	sd "github.com/aws/aws-sdk-go/service/servicediscovery"
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/pkg/apis/externaldns"
+	"github.com/kubernetes-incubator/external-dns/plan"
+	"github.com/linki/instrumented_http"
+	log "github.com/sirupsen/logrus"
+	"reflect"
+)
+
+const (
+	sdElbHostnameSuffix  = ".elb.amazonaws.com"
+	sdDefaultRecordTTL   = 300
+	sdServiceDescription = "Managed by Kubernetes External DNS (<owner-id>)"
+)
+
+// AWSServiceDiscoveryAPI is the subset of the AWS Route53 Auto Naming API that we actually use. Add methods as required.
+// Signatures must match exactly. Taken from https://github.com/aws/aws-sdk-go/blob/master/service/servicediscovery/api.go
+type AWSServiceDiscoveryAPI interface {
+	CreateService(input *sd.CreateServiceInput) (*sd.CreateServiceOutput, error)
+	DeregisterInstance(input *sd.DeregisterInstanceInput) (*sd.DeregisterInstanceOutput, error)
+	GetService(input *sd.GetServiceInput) (*sd.GetServiceOutput, error)
+	ListInstancesPages(input *sd.ListInstancesInput, fn func(*sd.ListInstancesOutput, bool) bool) error
+	ListNamespacesPages(input *sd.ListNamespacesInput, fn func(*sd.ListNamespacesOutput, bool) bool) error
+	ListServicesPages(input *sd.ListServicesInput, fn func(*sd.ListServicesOutput, bool) bool) error
+	RegisterInstance(input *sd.RegisterInstanceInput) (*sd.RegisterInstanceOutput, error)
+	UpdateService(input *sd.UpdateServiceInput) (*sd.UpdateServiceOutput, error)
+}
+
+// AWSSDProvider is an implementation of Provider for AWS Route53 Auto Naming.
+type AWSSDProvider struct {
+	client AWSServiceDiscoveryAPI
+	dryRun bool
+	// only consider namespaces ending in this suffix
+	namespaceFilter DomainFilter
+	// filter namespace by type (private or public)
+	namespaceTypeFilter *sd.NamespaceFilter
+	// refers to the owner id of the managed services
+	ownerID string
```

Comment:
Here we should not only care about ownership, but also about resources owning DNS records (or instances in case of SD). Which is also included in TXT records.

I will help you with this one, I will create a PR against your fork with my ideas how to solve this